types/opt/value.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package opt
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// Value is an optional value to be JSON-encoded.
+// Unlike the specialized Bool type, an empty (unset) Value[T] is marshaled as
+// a JSON null regardless of the "omitempty" JSON tag option when using `encoding/json`.
+// With https://github.com/go-json-experiment/json, however, empty optional Values are
+// omitted from the output.
+type Value[T any] struct {
+	value T
+	set   bool
+	isNil bool // indicates whether value is set to a nil slice or map
+}
+
+// Equal reports whether the receiver and the other value are equal.
+// If the template type T in Value[T] implements an Equal method, it will be used
+// instead of the == operator for comparing values.
+type equatable[T any] interface {
+	// Equal reports whether the receiver and the other values are equal.
+	Equal(other T) bool
+}
+
+var (
+	nilSliceJSON = []byte("[]")
+	nilMapJSON   = []byte("{}")
+)
+
+// Empty returns an empty optional Value.
+func Empty[T any]() Value[T] {
+	return Value[T]{set: false}
+}
+
+// Of returns an optional Value containing the specified value.
+// It treats nil pointers, unsafe pointers, and interfaces as unset values,
+// but nil slices and maps as empty slices and maps.
+func Of[T any](v T) Value[T] {
+	switch rv := reflect.ValueOf(v); rv.Kind() {
+	case reflect.Invalid:
+		// A nil interface (e.g. `any(nil)`) is equivalent to an unset value.
+		// It should be marshalled as null, or be omitted from the output.
+		return Empty[T]()
+	case reflect.Pointer, reflect.UnsafePointer, reflect.Interface:
+		if rv.IsNil() {
+			// A nil pointer, unsafe pointer, or interface is equivalent to an unset value.
+			// It should be marshalled as null, or be omitted from the output.
+			return Empty[T]()
+		}
+	case reflect.Map, reflect.Slice:
+		if rv.IsNil() {
+			// A nil map or slice is equivalent to an empty map or slice,
+			// and should be marshalled as `{}` or `[]`, accordingly.
+			return Value[T]{value: v, set: true, isNil: true}
+		}
+	}
+	return Value[T]{value: v, set: true, isNil: false}
+}
+
+// String implements Stringer.
+func (o *Value[T]) String() string {
+	if !o.set {
+		return fmt.Sprintf("(empty[%T])", o.value)
+	}
+	return fmt.Sprint(o.value)
+}
+
+// Set assigns the specified value to the optional value o.
+func (o *Value[T]) Set(v T) {
+	*o = Of(v)
+}
+
+// Clear resets o to an empty state.
+func (o *Value[T]) Clear() {
+	*o = Empty[T]()
+}
+
+// Present reports whether o has a value set.
+func (o *Value[T]) Present() bool {
+	return o.set
+}
+
+// Get returns the value and a flag indicating whether the value is set.
+func (o *Value[T]) Get() (v T, ok bool) {
+	return o.value, o.set
+}
```

Comment:
should these have value receivers instead?

Reply:
Done.